### PR TITLE
Fix/pull apply the content

### PR DESCRIPTION
**Pull request by petenelson**
### Description

Make a REST request for rendered post content after pulling a post from a network connection.

### Benefits

Allows the pulled post to get rendered shortcodes and anything else updated by `the_content` filter.

### Possible Drawbacks

The additional HTTP request has a small performance impact.

### Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/10up/distributor/blob/develop/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.

### Applicable Issues

#497

### CODE REVIEW

**Comment by jeffpaul** [comment]
@dkotter as you have availability, it would be great to get this through review so we can make any necessary changes and get this to a merge (and potential Distributor minor release to resolve this).

File: includes/classes/InternalConnections/NetworkSiteConnection.php [hunk 1]
```diff
@@ -267,6 +267,12 @@ public function pull( $items ) {
 
 			restore_current_blog();
 
+			// Allow the sync'ed post to be updated via a REST request to
```
Comment (dkotter) [inline includes/classes/InternalConnections/NetworkSiteConnection.php]: @petenelson If you could change these code comments into an actual DocBlock, similar to how most of the other actions are documented (and that should follow https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters), that would be great

File: includes/classes/InternalConnections/NetworkSiteConnection.php [hunk 2]
```diff
@@ -934,4 +940,53 @@ public static function wpseo_og_url( $og_url ) {
...
+		$rest_url = false;
+		if ( ! empty( $original_blog_id ) && ! empty( $original_post_id ) ) {
+			$rest_url = Utils\get_rest_url( $original_blog_id, $original_post_id );
+		}
+
+		if ( empty( $rest_url ) ) {
+			return;
+		}
+
+		$request = apply_filters( 'dt_update_content_via_request_args', [], $new_post_id, $this );
```
Comment (dkotter) [inline includes/classes/InternalConnections/NetworkSiteConnection.php:1009]: Same here, if we can add DocBlock here, that would be great

File: includes/classes/InternalConnections/NetworkSiteConnection.php [hunk 3]
```diff
@@ -934,4 +940,53 @@ public static function wpseo_og_url( $og_url ) {
...
+
+		$response = wp_remote_get( $rest_url, $request );
```
Comment (dkotter) [inline includes/classes/InternalConnections/NetworkSiteConnection.php]: To support VIP, I think we'll need to enhance this a bit. Basically we need to check if we are on WordPress.com VIP and if so, use the `vip_safe_wp_remote_get` function. Can see examples of this here: https://github.com/10up/distributor/blob/develop/includes/classes/ExternalConnections/WordPressExternalConnection.php#L184

File: includes/classes/InternalConnections/NetworkSiteConnection.php [hunk 4]
```diff
@@ -934,4 +940,53 @@ public static function wpseo_og_url( $og_url ) {
...
+
+		$body = wp_remote_retrieve_body( $response );
```
Comment (dkotter) [inline includes/classes/InternalConnections/NetworkSiteConnection.php]: We'll want to add a few more checks here, just to be safe. Basically check the `$response` value isn't a WP Error and check that the response code (`wp_remote_retrieve_response_code`) matches what we want. Can see examples of that here: https://github.com/10up/distributor/blob/develop/includes/classes/ExternalConnections/WordPressExternalConnection.php#L200

File: includes/utils.php [hunk 5]
```diff
@@ -783,3 +783,31 @@ function get_processed_content( $post_content ) {
...
+
+	$obj       = get_post_type_object( $post->post_type );
+	$rest_base = ! empty( $obj->rest_base ) ? $obj->rest_base : $obj->name;
+	$base      = sprintf( '%s/%s', 'wp/v2', $rest_base );
+
+	$rest_url = rest_url( trailingslashit( $base ) . $post->ID );
+
+	restore_current_blog();
+
+	return apply_filters( 'dt_get_rest_url', $rest_url, $blog_id, $post_id );
```
Comment (dkotter) [inline includes/utils.php:876]: Should have a DocBlock on this filter as well

File: includes/classes/InternalConnections/NetworkSiteConnection.php [hunk 6]
```diff
@@ -267,6 +267,12 @@ public function pull( $items ) {
 
 			restore_current_blog();
 
+			// Allow the sync'ed post to be updated via a REST request to
+			// get the rendered content.
+			if ( apply_filters( 'dt_pull_post_apply_rendered_content', true, $new_post_id, $this, $post_array ) ) {
```
Comment (dkotter) [inline includes/classes/InternalConnections/NetworkSiteConnection.php]: Not a CR comment but wondering if we want to default this to `false`? So sites that need the extra step of pulling in rendered content can opt-in but this won't impact other setups that don't need this? cc/ @jeffpaul

Comment (petenelson) [reply includes/classes/InternalConnections/NetworkSiteConnection.php]: I was thinking the same thing, with an eventual option in the settings to turn that off without a code change.

Comment (jeffpaul) [reply includes/classes/InternalConnections/NetworkSiteConnection.php]: My only concern is we're currently doing this for pushed posts so I imagine the perception is that we'd do this by default for pulled posts as well.  I _think_ I'd rather support the broader case and allow sites that know they don't need to render content first before pulling could adjust the default to `false` and gain back whatever performance the HTTP call is impacting.  Unless I'm misunderstanding things or not fully thinking through other scenarios, if so please call those out.

File: includes/classes/InternalConnections/NetworkSiteConnection.php (hunk 4 above)
Comment (petenelson) [reply includes/classes/InternalConnections/NetworkSiteConnection.php]: FYI, `wp_remote_retrieve_body()` already checks for WP_Error.

Comment (dkotter) [reply includes/classes/InternalConnections/NetworkSiteConnection.php]: Ah, didn't realize that. I still like the explicit check, so we can output a different error message in each instance, to better help with debugging issues.

**Review by dkotter — DISMISSED** [review]

**Comment by jeffpaul** [comment]
@petenelson looks like there's a merge conflict on this, could you help resolve that so we can look to merge this in?

**Comment by petenelson** [comment]
Merge conflict has been resolved

**Review by dkotter — APPROVED** [review]

**Pull request merged**